include/boost/graph/rank_spanning_branchings.hpp

```diff
@@ -0,0 +1,1159 @@
+            return true;
+          else
+            return false;
+        }
+
+     private:
+       const EdgeSet * p_es;
+
+    };
+
+    template<typename Edge, typename WeightMap, typename Compare>
+    struct BranchingEntry
+    {
+      Edge                                                        edge;
+      typename property_traits<WeightMap>::value_type             weight;
+      Compare                                                     compare;
+      unordered_set<Edge>                                         branching;
+      unordered_set<Edge>                                         include_edges;
+      unordered_set<Edge>                                         exclude_edges;
+      BranchingEntry(
+	const typename property_traits<WeightMap>::value_type& w,
+	const Edge& e,
+	Compare& comp,
+	const unordered_set<Edge>& b,
+	const unordered_set<Edge>& include,
+	const unordered_set<Edge>& exclude
+      ) :
+	  edge( e ), weight( w ), compare( comp ),
+	  branching( b ), include_edges( include ),
+	  exclude_edges( exclude ){}
+      bool operator<( BranchingEntry const & rhs ) const
+      { return compare( weight, rhs.weight ); }
+    };
+
+    template<typename WeightMap, typename Edge>
+    typename property_traits<WeightMap>::value_type
+    compute_branching_weight(
+      WeightMap& w,
+      const unordered_set<Edge>& branching
+    )
+    {
+
+       typedef typename property_traits<WeightMap>::value_type weight_t;
+
+       scoped_ptr<weight_t> weight;
+       weight.reset();
+
+       BOOST_FOREACH( const Edge& e, branching )
+       {
+         
+         if( !weight )
+         {
+           weight.reset( new weight_t );
+           *weight = get( w, e );
+         }
+         else
+         {
+	   *weight += get( w, e );
+         }
+       }
+
+       return *weight;
+
+    }
+     
+    template <typename Graph, typename BranchingProcessor, typename IndexMap,
+              typename WeightMap, typename Compare, typename Rank,
+              typename Parent>
+    void 
+    rank_spanning_branchings_impl( const Graph& g,
+			           BranchingProcessor bp,
+                                   IndexMap v_id,
+                                   WeightMap w,
+			           Compare comp,
+                                   Rank rank,
+                                   Parent pred1,
+                                   Parent pred2
+			         )
+    {
+
+      typedef typename graph_traits<Graph>::vertex_descriptor Vertex;
+      typedef typename graph_traits<Graph>::edge_descriptor Edge;
+
+      typedef typename property_traits<WeightMap>::value_type weight_t;
+
+      BOOST_CONCEPT_ASSERT(( VertexAndEdgeListGraphConcept<Graph> ));
+
+      BOOST_CONCEPT_ASSERT(( ReadablePropertyMapConcept<IndexMap, Vertex> ));
+      BOOST_CONCEPT_ASSERT(( ReadablePropertyMapConcept<WeightMap, Edge> ));
+
+      unordered_set<Edge> best_branching, empty_set;
+
+      shared_ptr< std::pair<Edge, weight_t> > p;
```

Comment:
Why `shared_ptr`? Is that really necessary?

Reply:
Like any programming construct they are not bad intrinsically, but they are often misused when a simpler construct would suffice.

Reply:
Thanks for this.  I used pointers to check for uninitialized or not found objects, but I think maybe boost::optional is better.  For example,

```suggestion
      boost::optional< std::pair<Edge, weight_t> > p;
```

The NEXT routine could then return the appropriate pair, if found.  If not found, p will be null (I guess really boost::none), as in the current implementation.  I have a branch that does that successfully.  Are there any caveats to boost::optional?